In [79]:
# Import libraries

import pandas as pd 
import numpy as np

import matplotlib as plt
import seaborn as sns 


In [92]:
# Here I load the Sheffield Collision Dataset

df = pd.read_csv("../data/Filtered_Sheffield_Traffic_Data.csv")

print(df.dtypes)

collision_index                                      object
collision_year                                        int64
collision_ref_no                                     object
location_easting_osgr                               float64
location_northing_osgr                              float64
longitude                                           float64
latitude                                            float64
police_force                                          int64
collision_severity                                    int64
number_of_vehicles                                    int64
number_of_casualties                                  int64
date                                                 object
day_of_week                                           int64
time                                                 object
local_authority_district                              int64
local_authority_ons_district                         object
local_authority_highway                 

In [81]:
# Here I display the first 5 rows

df.head(5)

,collision_index,collision_year,collision_ref_no,location_easting_osgr,location_northing_osgr,longitude,latitude,police_force,collision_severity,number_of_vehicles,...,carriageway_hazards_historic,carriageway_hazards,urban_or_rural_area,did_police_officer_attend_scene_of_accident,trunk_road_flag,lsoa_of_accident_location,enhanced_severity_collision,collision_injury_based,collision_adjusted_severity_serious,collision_adjusted_severity_slight
0,200514D151505,2005,14D151505,435910.0,389680.0,-1.461345,53.402715,14,2,1,...,0,0,1,1,2,E01007869,-1,0,1.000000,0.000000
1,199914HH00487,1999,14HH00487,443640.0,381260.0,-1.346245,53.326453,14,3,3,...,0,0,1,-1,2,E01008023,-1,-1,NaN,NaN
2,200714G118607,2007,14G118607,435830.0,387460.0,-1.462800,53.382766,14,3,1,...,2,17,1,1,2,E01007884,-1,0,0.248913,0.751087
3,200214KK01181,2002,14KK01181,439370.0,387820.0,-1.409537,53.385751,14,3,1,...,0,0,1,-1,2,E01007907,-1,-1,NaN,NaN
4,200414EE00565,2004,14EE00565,444030.0,381380.0,-1.340373,53.327500,14,2,2,...,0,0,1,-1,2,E01008019,-1,0,1.000000,0.000000


In [82]:
# Here I check the dataset information

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7933 entries, 0 to 7932
Data columns (total 44 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   collision_index                                   7933 non-null   object 
 1   collision_year                                    7933 non-null   int64  
 2   collision_ref_no                                  7933 non-null   object 
 3   location_easting_osgr                             7868 non-null   float64
 4   location_northing_osgr                            7868 non-null   float64
 5   longitude                                         3693 non-null   float64
 6   latitude                                          3693 non-null   float64
 7   police_force                                      7933 non-null   int64  
 8   collision_severity                                7933 non-null   int64  
 9   number_of_vehicles 

In [83]:
# Check dataset size

print(df.shape)

(7933, 44)


In [84]:
# Checking columns

print(df.columns)

Index(['collision_index', 'collision_year', 'collision_ref_no',
       'location_easting_osgr', 'location_northing_osgr', 'longitude',
       'latitude', 'police_force', 'collision_severity', 'number_of_vehicles',
       'number_of_casualties', 'date', 'day_of_week', 'time',
       'local_authority_district', 'local_authority_ons_district',
       'local_authority_highway', 'local_authority_highway_current',
       'first_road_class', 'first_road_number', 'road_type', 'speed_limit',
       'junction_detail_historic', 'junction_detail', 'junction_control',
       'second_road_class', 'second_road_number',
       'pedestrian_crossing_human_control_historic',
       'pedestrian_crossing_physical_facilities_historic',
       'pedestrian_crossing', 'light_conditions', 'weather_conditions',
       'road_surface_conditions', 'special_conditions_at_site',
       'carriageway_hazards_historic', 'carriageway_hazards',
       'urban_or_rural_area', 'did_police_officer_attend_scene_of_accident',
 

In [85]:
# Here I start checking for missing values in dataset 

missing_values = df.isnull().sum()

print(missing_values )


collision_index                                        0
collision_year                                         0
collision_ref_no                                       0
location_easting_osgr                                 65
location_northing_osgr                                65
longitude                                           4240
latitude                                            4240
police_force                                           0
collision_severity                                     0
number_of_vehicles                                     0
number_of_casualties                                   0
date                                                   0
day_of_week                                            0
time                                                   0
local_authority_district                               0
local_authority_ons_district                           0
local_authority_highway                                0
local_authority_highway_current

In [86]:
# Calculate the total number of cells with missing values

total_missing_values = missing_values.sum()

print("Total Missing value:", total_missing_values)

Total Missing value: 24398


In [87]:
# Here I calculate the total cells in dataset

total_columns = df.shape[1]

total_rows = df.shape[0]

total_cells = total_rows * total_columns 

print(total_cells)

349052


In [88]:
overall_missing_percentage = ( total_missing_values / total_cells ) * 100

print(f"Overall Missing Percentage: {overall_missing_percentage:.2f}%")

Overall Missing Percentage: 6.99%


In [89]:
# This accounts for the percentage of data missing from the columns with missing values

print(missing_values["location_northing_osgr"] / total_rows * 100)

print(missing_values["location_easting_osgr"] / total_rows * 100)


0.8193621580738686
0.8193621580738686


In [90]:
missing_values[missing_values > 0]

for column in missing_values:
    column = missing_values / total_rows * 100

print(column[column > 0])


location_easting_osgr                   0.819362
location_northing_osgr                  0.819362
longitude                              53.447624
latitude                               53.447624
local_authority_highway_current        66.986008
collision_adjusted_severity_serious    66.015379
collision_adjusted_severity_slight     66.015379
dtype: float64


In [91]:
duplicate_row = df.duplicated().sum()

print("Duplicated row:", duplicate_row)

Duplicated row: 0


## Missing value analysis

The dataset conains missing values in several columns

The highest number of missing values were found in: 

- local_authority_highway_current: 5314
- collision_adjusted_severity_serious: 5237
- collision_adjusted_severity_slight: 5237

The longitude and latitude columns also contain a lot of missing values 4240 each. 

The location northing and easting contain only 65 missing values, which represents less than 1% of the dataset

## Duplicate value analysis

I checked the dataset for any duplicate values, however none were found. 